### PR TITLE
Fixes #3615: Add author, title, publishDate fields to '/' route

### LINKED ISSUE #3615: [Post-services] Add `author`, `title`, `publishDate` fields to '/' route
Our current Post-services `/` route will return an array of posts see JSON below.
```json
[
  {
    "id": "33a8c63f31",
    "url": "https://api.telescope.cdot.systems/v1/posts/33a8c63f31"
  },
  {
    "id": "6f3d109b82",
    "url": "https://api.telescope.cdot.systems/v1/posts/6f3d109b82"
  },
  {
    "id": "3253659d90",
    "url": "https://api.telescope.cdot.systems/v1/posts/3253659d90"
  }
  ]
  ```
 Since our React native app design uses a click and reads design. When fetching the list of posts, we have to do another fetch to get `author`, `title`, `date`. 
 This is inefficient since it makes a lot of unnecessary requests, we only need to request post detail on click to a post.
image

The solution is to add `author`, `title`, and `publishDate` to the object.
Sample JSON:
```json
[
  {
    "id": "33a8c63f31",
    "url": "https://api.telescope.cdot.systems/v1/posts/33a8c63f31",
    "author":"",
    "title":"",
    "publishDate":""
  },
  {
    "id": "6f3d109b82",
    "url": "https://api.telescope.cdot.systems/v1/posts/6f3d109b82",
     "author":"",
    "title":"",
    "publishDate":""
  },
  {
    "id": "3253659d90",
    "url": "https://api.telescope.cdot.systems/v1/posts/3253659d90",
     "author":"",
    "title":"",
    "publishDate":""
  }
  ]
  ```

### PR DESCRIPTION
**Pull request by liutng**
## Issue This PR Addresses
Closes #3615 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
In this PR, an extra query param `expand` is added to the URL /posts in posts services. 
When `expand=1`, there will be extra data of author, title, and publishDate returned.
When expand does not equal 1, or expand does not present in the request, the request will return the data like before.

## Steps to test the PR
Load the posts page in the react-native app, and observe if the extra fields of data are returned by `/posts`. This requires the react-native app to add "expand=1" to its request query params of `/posts` as well.

Here's what does the request and data look like.
image

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by P-DR0ZD — APPROVED** [review]
Looks Good

**Comment by liutng** [comment]
> Looks Good

Thanks.

File: src/api/posts/test/posts.test.js [hunk 1]
```diff
@@ -69,6 +69,19 @@ describe('/posts', () => {
     expect(res.body.length).toBeGreaterThan(0);
   });
 
+  test('request posts with a valid expand query param', async () => {
```
Comment (humphd) [inline src/api/posts/test/posts.test.js:72]: Please add a test for `expand=1` as well.

File: src/api/posts/src/routes/posts.js [hunk 2]
```diff
@@ -60,14 +60,26 @@ posts.get('/', validatePostsQuery(), async (req, res, next) => {
     first: `/posts?per_page=${perPage}&page=${1}`,
     last: `/posts?per_page=${perPage}&page=${Math.floor(postsCount / perPage)}`,
   });
-  res.json(
-    ids
+  let data;
+  if (expand === 1) {
+    data = await Promise.all(
```
Comment (humphd) [inline src/api/posts/src/routes/posts.js:65]: This can fail, needs `try/catch`

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/src/routes/posts.js (hunk 2 above)
Comment (liutng) [reply src/api/posts/src/routes/posts.js:65]: Thank for the suggestion, a `try/catch` block is now added to wrap `Promise.all()`.

File: src/api/posts/test/posts.test.js (hunk 1 above)
Comment (liutng) [reply src/api/posts/test/posts.test.js:72]: I tried to cover this scenario in my test but it either returns empty data or timeout, although the API works fine if I were requesting with my browser or Postman. Got any clue what might cause this problem?

File: src/api/posts/src/routes/posts.js (hunk 2 above)
Comment (humphd) [reply src/api/posts/src/routes/posts.js:65]: When you say "is now added," there is nothing here.  Please push your changes.

File: src/api/posts/test/posts.test.js (hunk 1 above)
Comment (humphd) [reply src/api/posts/test/posts.test.js:72]: Debug it, add logging, etc.  You should have 150 posts created before this test runs.  We can't land this change without a test that proves it works.

File: src/api/posts/src/routes/posts.js (hunk 2 above)
Comment (liutng) [reply src/api/posts/src/routes/posts.js]: Sorry, I forgot to push my code, it's now pushed.

**Review by humphd — COMMENTED** [review]
Looks like the test changes are still pending?  What's happening with those?

**Comment by liutng** [comment]
> Looks like the test changes are still pending? What's happening with those?

I'm still working on the test where `expand=1`.

**Comment by liutng** [comment]
Finally, the test is working. I added extra feed data to `posts.test.js` to make it work. The code has been pushed now.

File: src/api/posts/test/posts.test.js [hunk 3]
```diff
@@ -69,6 +86,25 @@ describe('/posts', () => {
...
+    const res = await request(app).get('/?page=1&expand=0');
+    expect(res.status).toEqual(200);
+    // This will depend on the env value, so as long as we get back something.
+    expect(res.body.length).toBeGreaterThan(0);
+  });
+  test('request posts with a valid expand query param', async () => {
+    const res = await request(app).get('/?page=1&expand=1');
+    expect(res.status).toEqual(200);
+    // This will depend on the env value, so as long as we get back something.
+    expect(res.body.length).toBeGreaterThan(0);
```
Comment (humphd) [inline src/api/posts/test/posts.test.js:99]: This isn't enough.  In the case that you include `expand=1`, we'll expect the shape of the objects returned in the array to be different.  Please add some logic to the tests to deal with the case of `expand=1` and without, looking at the properties in the objects.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/test/posts.test.js (hunk 3 above)
Comment (liutng) [reply src/api/posts/test/posts.test.js:99]: Got it, I've pushed a revised test that now checks the extra properties in the return data when `expand=1`.

File: src/api/posts/src/routes/posts.js [hunk 4]
```diff
@@ -60,14 +60,31 @@ posts.get('/', validatePostsQuery(), async (req, res, next) => {
...
+            publishDate: post.published,
+          };
+        })
+      );
+    } catch (err) {
+      logger.error({ err }, 'Unable to get detail information of posts from Redis');
+      next(createError(503, 'Unable to connect to database'));
+    }
+  } else
+    data = ids
```
Comment (Kevan-Y) [inline src/api/posts/src/routes/posts.js:84]:
```suggestion
const data = ids.map((id) => ({ id, url: `${postsUrl}/${id}` }));

  if (expand === 1) {
    try {
      const expandedData = await Promise.all(
        data.map(async ({id, url}) => {
          // obtain the corresponding feed to populate the author's name to the return data.
          const post = await Post.byId(id);
          return {
            id,
            url,
            author: post.feed.author,
            title: post.title,
            publishDate: post.published,
          };
        })
      );
      
      res.json(expandedData);
    } catch (err) {
      logger.error({ err }, 'Unable to get detail information of posts from Redis');
      next(createError(503, 'Unable to connect to database'));
    }
  }
  
   res.json(data);
```

Suggestion: I think you could simplify like this to not repeats some logic

Comment (liutng) [reply src/api/posts/src/routes/posts.js:84]: Thank you very much. This would reduce our code's repetition. However, after the change, ``ids.map((id) => ({ id, url: `${postsUrl}/${id}` }));`` will always be executed regardless the expand parameter value. This could potentially slow down our code and it would be better if we keep it as it was.

Comment (Kevan-Y) [reply src/api/posts/src/routes/posts.js:84]: Yes, but you will need this data in both case anyway, also we are not processing millions of data, which I believe it won't slow down performance.

Comment (liutng) [reply src/api/posts/src/routes/posts.js:84]: I tried your solution and the result shows that when `expand=1` the posts services will crash since ` res.json(expandedData);` already sent back a json to the client and `res.json(data);` will use this already-consumed res object to send back a response to the client again. The posts service log looks like this: 
```
2022-11-19 18:40:12 Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
2022-11-19 18:40:12     at new NodeError (node:internal/errors:387:5)
2022-11-19 18:40:12     at ServerResponse.setHeader (node:_http_outgoing:644:11)
2022-11-19 18:40:12     at ServerResponse.header (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:794:10)
2022-11-19 18:40:12     at ServerResponse.send (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:174:12)
2022-11-19 18:40:12     at ServerResponse.json (/app/node_modules/.pnpm/express@4.18.2/node_modules/express/lib/response.js:278:15)
2022-11-19 18:40:12     at /app/src/routes/posts.js:87:7
2022-11-19 18:40:12     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
2022-11-19 18:40:12   code: 'ERR_HTTP_HEADERS_SENT'
2022-11-19 18:40:12 }
```

Comment (Kevan-Y) [reply src/api/posts/src/routes/posts.js:84]: Have you try to put `return res.json(expendedDate)`

Comment (liutng) [reply src/api/posts/src/routes/posts.js:84]: I just tried `return res.json(expendedDate)` and it works now. Thank you for your suggestion.

File: src/api/posts/src/routes/posts.js [hunk 5]
```diff
@@ -60,14 +60,31 @@ posts.get('/', validatePostsQuery(), async (req, res, next) => {
     first: `/posts?per_page=${perPage}&page=${1}`,
     last: `/posts?per_page=${perPage}&page=${Math.floor(postsCount / perPage)}`,
   });
-  res.json(
-    ids
+  let data = {};
```
Comment (humphd) [inline src/api/posts/src/routes/posts.js:63]: nit: don't bother assigning a value here, since you'll throw it away in the if/else that follows.  Another pattern you could use is to move the if/else to a function:

```js
try {
  const data = await prepareData(ids, expand === 1);
  res.json(data);
} ... 
```

And your function could look like this (untested):

```js
const prepareData = async (ids, shouldExpand) {
  if (shouldExpand) {
      return Promise.all(
        ids.map(async (id) => {
          // obtain the corresponding feed to populate the author's name to the return data.
          const post = await Post.byId(id);
          return {
            id,
            url: `${postsUrl}/${id}`,
            author: post.feed.author,
            title: post.title,
            publishDate: post.published,
          };
        })
      );
  } else
    return ids
      .map((id) => ({
        id,
        url: `${postsUrl}/${id}`,
      }))
  );
};
```

File: src/api/posts/test/posts.test.js [hunk 6]
```diff
@@ -11,7 +11,21 @@ describe('/posts', () => {
...
+    const id = hash(guid);
+    return {
+      id,
+      author: 'author',
+      url: 'http://foo.url',
+      user: 'user',
+      link: 'http://foo.link',
+      etag: 'etag',
+      lastModified: new Date('2009-09-07T22:23:00.544Z'),
+      githubUsername: 'githubusr',
```
Comment (humphd) [inline src/api/posts/test/posts.test.js:25]: s/githubusr/githubUsername/ to make it even more clear.

File: src/api/posts/test/posts.test.js (hunk 3 above)
Comment (humphd) [reply src/api/posts/test/posts.test.js:99]: This is getting better, but still doesn't test what we need.

Use Jest's [`arrayContaining`](https://jestjs.io/docs/expect#expectarraycontainingarray) and [`objectContaining`](https://jestjs.io/docs/expect#expectobjectcontainingobject) to look for one item that has the shape/values you expect in the results returned.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/test/posts.test.js (hunk 6 above)
Comment (liutng) [reply src/api/posts/test/posts.test.js:26]: Sorry, what does this mean? Should I change the field name of `githubUsername` to something else? Or the value `githubusr`?

Comment (humphd) [reply src/api/posts/test/posts.test.js:26]: The value.  It's easiest if the property name and value match (e.g., like `etag` above).

Comment (liutng) [reply src/api/posts/test/posts.test.js]: Got it, thank you for the suggestion, it's already fixed now and pushed.

File: src/api/posts/src/routes/posts.js [hunk 7]
```diff
@@ -7,11 +7,35 @@ const postsUrl = process.env.POSTS_URL || '/';
 
 const posts = Router();
 
+const prepareData = async (ids, shouldExpand = false) => {
+  if (shouldExpand) {
+    const data = await Promise.all(
```
Comment (humphd) [inline src/api/posts/src/routes/posts.js:12]: No need for `data` or `await` here, just return the Promise directly:

```js
return Promise.all(
```

**Review by humphd — CHANGES_REQUESTED** [review]
Nice, almost there.

File: src/api/posts/src/routes/posts.js (hunk 7 above)
Comment (liutng) [reply src/api/posts/src/routes/posts.js]: Hi, I pushed a new commit.

File: src/api/posts/src/routes/posts.js [hunk 8]
```diff
@@ -9,7 +9,7 @@ const posts = Router();
 
 const prepareData = async (ids, shouldExpand = false) => {
```
Comment (humphd) [inline src/api/posts/src/routes/posts.js:10]: Get rid of `async` here, we aren't going to `await` anymore.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/src/routes/posts.js (hunk 8 above)
Comment (liutng) [reply src/api/posts/src/routes/posts.js:10]: Got it. A new commit is pushed.

**Review by humphd — APPROVED** [review]
This looks good to me, nice work.  Once this goes green on CI, please squash all 12 commits into 1, and rebase on latest `master`.

**Comment by liutng** [comment]
Ok, I will do it now.

**Comment by humphd** [comment]
This did not go to plan:

Screen Shot 2022-11-21 at 10 36 39 AM

You have merged master into your branch, which is never what you want to do when creating PRs (they will merge your branch into master).

Can you do this again correctly?  See https://telescope.cdot.systems/docs/contributing/git-workflow for help.

**Comment by liutng** [comment]
Oops.. I will try to figure out how to undo it and redo it.

**Comment by liutng** [comment]
I tried everything I could to fix this PR. Maybe this Pull Request is beyond repairable.. Should I close this one and open a new one?

**Comment by liutng** [comment]
Thank you, I will not hesitate to contact you when I need help and also the lesson I've learned from this time is never use any git command if I'm not aware of what this command does.
